utils.py

```diff
@@ -29,3 +32,65 @@ def deserialize_from_dill(dill_path: Any) -> Any:
     logging.error('Failed to deserialize %s: File does not exist: %s',
                   dill_path, e)
     return None
+
+
+def retryable(exceptions=None,
+              default_retries=5,
+              delay_fn=None,
+              other_exceptions=None,
+              default_return=None):
+  """
+    Decorator that retries the function on specified exceptions.
+    :param exceptions: List/Set of exceptions or a dictionary of exceptions with
+      custom retry counts.
+    :param default_retries: Number of retries if no custom count is provided.
+    :param delay_fn: Function to determine the delay between retries. Default
+      is random between 0-60 seconds.
+    :param default_return: The value returned if all retries fail.
+    """
+  exception_config = {
+      exc: default_retries for exc in exceptions or {}
+  } | (other_exceptions or {})
+  valid_exceptions = tuple(e for e in exception_config.keys()
+                           if isinstance(e, type) and issubclass(e, Exception))
+  if not valid_exceptions:
+    raise ValueError("No valid exception classes provided.")
+
+  def decorator(func):
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+      retry_count = 0
+      while retry_count < default_retries:
+        try:
+          return func(*args, **kwargs)
+        except valid_exceptions as e:  # pylint: disable=catching-non-exception
+          exc_name = type(e).__name__
+          retry_count += 1
+          logging.info(
+              'Retrying %s due to %s. Attempt %d with args=%s, kwargs=%s',
+              func.__name__, exc_name, retry_count, args, kwargs)
+
+          # Get the number of retries for this exception, or use the default
+          current_retries = exception_config.get(type(e), default_retries)
+          if delay_fn:
+            delay = delay_fn(exc_name, retry_count)
+          else:
+            delay = random.uniform(0, 60)
+          time.sleep(delay)
+
+          if retry_count >= current_retries:
+            logging.error('Max retries reached for %s. Exiting.', exc_name)
+            break
+        except Exception as e:
+          logging.error('Unhandled exception: %s. Exiting.', e)
```

Comment:
shouldn't we re-raise the exception? Or just not catch these? 

It seems dangerous to let things proceed if an exception happens that we don't expect.